### ISSUE TITLE
much more transaction info needed when reverted

### ISSUE THREAD

**Issue by henry-hz** [foundry-rs/foundry#3093]
### Component

Cast

### Describe the feature you would like

When a transaction fails, on cast we receive a really small and mystical information.  See the difference between seth and cast below. Maybe because the output does not include the transaction hash [so it becomes impossible to debug], this feature request could be considered a bug.

```
dex master* ❯ cast send 0x856e4424f806D16E8CBC702B3c0F2ede5468eae5 "claim(uint256)" 1                                                                                                                        22:45:44
Insert keystore password:

Error:
(code: 3, message: execution reverted, data: Some(String("0x4e487b710000000000000000000000000000000000000000000000000000000000000011")))

dex master* ❯ seth send 0x856e4424f806D16E8CBC702B3c0F2ede5468eae5 "claim(uint256)" 1                                                                                                                        22:46:01
seth-send: warning: `ETH_GAS' not set; using default gas amount
seth-send: Published transaction with 36 bytes of calldata.
seth-send: 0xcad400db3d7296437c396e1335baa3d15397f9deb9e209e95091eb2cb704b982
seth-send: Waiting for transaction receipt...
seth-send: Transaction failed.
blockHash               0x231c4d68988c27b7881ff1f537e471a6e386b545a969c67cbf8d4782449f3743
blockNumber             31
contractAddress         null
cumulativeGasUsed       104942
effectiveGasPrice       1020779290
from                    0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266
gasUsed                 104942
logs                    []
logsBloom               0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
status                  0
to                      0x856e4424f806d16e8cbc702b3c0f2ede5468eae5
transactionHash         0xcad400db3d7296437c396e1335baa3d15397f9deb9e209e95091eb2cb704b982
transactionIndex        0

```

**Comment by tgfukuda** [comment]
I think the main reason of such a small information comes from a failure in gas estimation.

Below is executed on mainnet fork and `0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2` is WETH9.
The sender is anvil's generated address and have no WETH9, so it would fail.
```
cast send --rpc-url http://127.0.0.1:8545 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2 'withdraw(uint)' $(cast --to-wei 1 eth)
(code: 3, message: execution reverted, data: Some(String("0x")))
```
anvil log:
```
eth_chainId
eth_getTransactionCount
eth_getBlockByNumber
eth_feeHistory
eth_estimateGas
```

If we provide gas constrants to `cast`, then it will send the given transaction though it may reverted.

```
cast send --rpc-url http://127.0.0.1:8545 --gas-limit 7000000 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2 'withdraw(uint)' $(target/debug/cast --to-wei 1 eth)

blockHash               0x1bac83e90ae264a34eaa874f26dc3114d2829bdcafc295ae187384d663fca97d
blockNumber             15570278
contractAddress         
cumulativeGasUsed       23731
effectiveGasPrice       11700454506
gasUsed                 23731
logs                    []
logsBloom               0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
root                    
status                  0
transactionHash         0x303f38372ebc5a27b3c5f6d524faeb8378aedaf82f44ea6dc9c369d612d13cf3
transactionIndex        0
type
```
anvil log:
```
eth_chainId
eth_getTransactionCount
eth_getBlockByNumber
eth_feeHistory
eth_sendRawTransaction

    Transaction: 0x303f38372ebc5a27b3c5f6d524faeb8378aedaf82f44ea6dc9c369d612d13cf3
    Gas used: 23731
    Error: reverted without a reason string

    Block Number: 15570278
    Block Hash: 0x1bac83e90ae264a34eaa874f26dc3114d2829bdcafc295ae187384d663fca97d
    Block Time: "Mon, 19 Sep 2022 21:51:24 +0000"

eth_getTransactionReceipt
```

However, it is a good idea to give some hints or set default gas like `seth`.
I think the best approach is using evm, but I'm not sure it is possible.

**Comment by microbecode** [comment]
I'm new to Foundry, but wondering about this weird error message `Some(String("0x"))`, even if there should be a real revert string. What could be done about this? Thanks

**Comment by nftchef** [comment]
It appears that the `Some(String("0x"))` message is appearing when using custom Error functions/reverts and returns the Error function selector. is there support for getting the string representation of the error?

**Comment by mattsse** [comment]
> It appears that the Some(String("0x")) message is appearing 

appearing where exactly?

decoding custom errors without knowing the abi is a bit hard. we have some fuzzy matching for custom errors somewhere but it don't think it's enabled here.

**Comment by nftchef** [comment]
sorry about that,

`Some(String("0x"))` is appearing when using `cast send ` on a function with a custom error revert. e.g.,

**Comment by mattsse** [comment]
looks like this is a bug in revm that has been fixed in 3.0 release but we haven't migrated yet.

ref #4356

**Comment by zerosnacks** [comment]
Marking as resolved due to https://github.com/foundry-rs/foundry/issues/4356 having been merged
